spec/lucky/param_serializable_spec.cr

```diff
@@ -0,0 +1,199 @@
+require "../spec_helper"
+
+include ContextHelper
+include MultipartHelper
+
+class BasicParams
+  include Lucky::ParamSerializable
+  skip_param_key
+
+  property string : String
+  property int16 : Int16
+  property int32 : Int32
+  property int64 : Int64
+  property bool : Bool
+  property float64 : Float64
+  property uuid : UUID
+  property blank : String?
+end
+
+class UserWithKeyParams
+  include Lucky::ParamSerializable
+  param_key :user
+
+  property name : String
+  property age : Int32
+  property fellowship : String?
+end
+
+class ComplexParams
+  include Lucky::ParamSerializable
+
+  property tags : Array(String)
+  property numbers : Array(Int32)
+  property default : Bool = true
+  @[Lucky::ParamField(param_key: :override)]
+  property version : Float64
+  @[Lucky::ParamField(ignore: true)]
+  property internal : Int32 = 4
+end
+
+class CrashingParams
+  include Lucky::ParamSerializable
+  skip_param_key
+
+  property required_but_missing : String
+  @[Lucky::ParamField(param_key: :key)]
+  property wrong : Bool
+end
+
+# {
+#   "query":{
+#     "bool":{
+#       "must":[
+#          {"terms":{"brand":["micromax","samsung"]}}
+#       ] ,
+#       "should":[
+#          { "range": { "price": { "gte": 6000, "lte": 10000 } } },
+#          { "range": { "price": { "gte": 16000, "lte": 30000 } } }
+#       ]
+#     }
+#   }
+# }
+
+# class SearchParams
+#   include Lucky::ParamSerializable
+
+#   property q : String
+#   property page : Int32
+#   property per : Int32 = 50
+#   property sort : Array(String)
+#   @[Lucky::ParamField(param_key: :filter)]
+#   property active : Bool = false
+#   @[Lucky::ParamField(param_key: :filter)]
+#   property city : String
+
+# end
+
+describe Lucky::ParamSerializable do
+  describe "param_key" do
+    it "checks the key on all params" do
+      request = build_request
+      request.query = "user:name=Gandalf&user:age=11000&fellowship=bracelet"
+
+      params = Lucky::Params.new(request)
+      user_params = UserWithKeyParams.from_params(params)
+
+      user_params.param_key.should eq "user"
```

Comment:
I'm not sure this is something that should be exposed. Preferrably, setting the param key of the class would essentially set a default param key for each attribute if it isn't specified. A different param key could be specified on every attribute and this wouldn't make sense.

Reply:
Good point. I can make that method private.